lib/ansible/executor/task_executor.py

```diff
@@ -750,10 +752,17 @@ def _evaluate_failed_when_result(result):
                 if C.INJECT_FACTS_AS_VARS:
                     variables.update(clean_facts(af))
 
-        # save the notification target in the result, if it was specified, as
+        # Save the notification target in the result, if it was specified, as
         # this task may be running in a loop in which case the notification
-        # may be item-specific, ie. "notify: service {{item}}"
-        if self._task.notify is not None:
+        # may be item-specific, ie. "notify: service {{item}}".
+        # Save all notification targets from all loop items if any item has
+        # changed and loop_control.notify_scope='task' (default) or save only
+        # notification targets only from items which changed if loop_control.notify_scope='per_loop_item'
+        if (
+                self._task.notify is not None and (
+                    notify_scope != 'per_loop_item' or (
+                        notify_scope == 'per_loop_item' and
+                        result['changed']))):
```

Comment:
This is probably just me but I had a hard time parsing that. I would change `notify_scope != 'per_loop_item'` into `notify_scope == 'task'` and put it all on the one or two lines maybe. Not a blocker, of course.

Reply:
This doesn't account for invalid values. I'm wondering if this would be better as a boolean value rather than a string field. Something like `notify_per_item` with a default of `False`.

Are there other values for this option that would make it better suited to a string than a boolean?

If we do keep it as a string value, maybe the options could be `task, item` or `task, per_item`. Just throwing out ideas.

Reply:
This is easier to read:
```python
if self._task.notify is not None:
    if notify_per_loop_item:
        if result['changed']:
            result['_ansible_notify'] = self._task.notify
    else:
        result['_ansible_notify'] = self._task.notify
```

Or using the existing string option:
```python
if self._task.notify is not None:
    if notify_scope not in ['task', 'per_loop_item']:
        raise AnsibleError("Invalid option '{0}' given for notify_scope. Valid options are "
                           "'task' or 'per_loop_item'".format(to_text(notify_scope)))
    elif notify_scope == 'per_loop_item':
        if result['changed']:
            result['_ansible_notify'] = self._task.notify
    else:
        result['_ansible_notify'] = self._task.notify
```

Reply:
> This doesn't account for invalid values.

Yeah, I realized that when suggested `notify_per_item` below, too ;-) I was wondering if we had anything like `choices` for `FieldAttribute` so we didn't have to check for invalid types here in the code but not that I could find.

Reply:
The way we have done this before is via something like `_validate_debugger`.  That method effectively does this, but for `debugger`

Reply:
Nice. So I'd go with `_validate_$whatever_name_we_bikeshed` method to check for invalid values. Unless we decide to use a boolean.